### ISSUE TITLE
UI Styles: Implement 'box-shadow'

### ISSUE THREAD

**Issue by bryphe** [revery-ui/revery#72]
For `box-shadow`, we'll want to add a `Style` property, something like:
```
Style.make(~boxShadow=BoxShadow(-5., -5., 10., 10., Color.rgba(0., 0., 0., 0.5), ..., ())
```

The box-shadow properties would mirror the properties from CSS: https://developer.mozilla.org/en-US/docs/Web/CSS/box-shadow

(in order)
- __xOffset__ - the horizontal offset of the shadow
- __yOffset__ - the vertical offset of the shadow
- __blur radius__
- __spread radius__
- __color__

Note that the box shadow won't impact Layout, so we don't need to worry about passing it to layout (like we did for flex properties!)

The way I would think about tackling this is splitting it up into a few parts:

__Part 1: Set up the types__

For part 1 - I'd just look at adding the types in `Style.re`, and plumbing them through.

We'd need to:
- Add a `BoxShadow` type
- Add a `boxShadow` property to the Style record / make fucntions

__Part 2: Render a solid shadow__

For part 2, I'd skip the blur / spread to simplify things - we can just render a quad the same size as the view node, but offset based on the values of `xOffset` and `yOffset`.

The place we'd want to look at drawing this is in the `ViewNode` - this is the code that draws the `background` of a node: 
https://github.com/bryphe/revery/blob/4a52028176aad32f81973a3a0bbc651299359dda/src/UI/ViewNode.re#L48

Prior to drawing that - we'd want to draw the 'shadow' quad. Right now, our rendering looks like this:
- Set `uWorld` for background quad - this is the transform matrix
- Set `uColor` for background quad - this sets the color in the shader

With our shadow, we'd want to do this in two passes:
- Render shadow
  - Set `uWorld` for shadow - this is like the transform for the background quad, but with an extra translation to set the xOffset/yOffset.
  - Set `uColor` for the shadow - this would be based off the shadow color
  - Draw via `Geometry.draw(_quad, solidShader)`
- Render background
  - Set `uWorld` for background (same as today)
  - Set `uColor` for background (same as today)
  - Draw via `Geometry.draw(_quad, solidShader)`

So our shadow would add an extra draw call

__Part 3: Render the blurred edges__

For part 3, we'd take into account the `spread` and `blurRadius`.

I think the easiest way to handle this (at least in a quick way) would be to add __additional quads__ for the edges, and render those edges with a shader that handles the gradient. At the 'core' of our shadow, the opacity would be `1.0` x the shadow color. At the edge of the shadow, it'll fall off to `0` - becoming more transparent as it spreads out. We could use additional quads around the core shadow, along with a shader that models this fall-off to get a smooth gradient. 

There might be better ways to handle this, too.

**Comment by Akin909** [comment]
I'll have a look at this @bryphe, any tips?

**Comment by Akin909** [comment]
@bryphe re. setting the `uWorld` for the shadow does that mean creating a similar function like `getWorldTransform` in `Node.re` ?

**Comment by bryphe** [comment]
> @bryphe re. setting the uWorld for the shadow does that mean creating a similar function like getWorldTransform in Node.re ?

Essentially we'd have to do something like this: 
```
let shadowTransform = Mat4.create();
Mat4.fromTranslation(shadowTransform, Vec3.create(shadowOffsetX, shadowOffsetY));

let worldShadowTransform = Mat4.create();
worldShadowTransform = Mat4.multiply(worldShadowTransform, _this#getWorldTransform(), shadowTransform);
```

And then we'd set `uWorld` to do that. The `worldTransform` is all of the parent transforms applied (any animations, offsets, etc), and then the `shadowTransform` is just the transform we need for the shadow. Essentially what this is saying is 'First, move my shadow quad based on the shadowOffsetX, shadowOffsetY. Then, apply each parent transform in turn'. Parent transforms might include rotations, scales, etc - they would all get applied in this case.

Hope that helps! Reading through the [transforms](https://learnopengl.com/Getting-started/Transformations) section here might help too

**Comment by Akin909** [comment]
@bryphe thanks that helped a lot :+1:,got another question now I've actually drawn the shadow, I'm not sure I followed the strategy you suggested about  using more quads with lessening opacity to draw the blur?

**Comment by OhadRau** [comment]
What are your thoughts on named parameters for the BoxShadow? Right now I think it's a little confusing if you don't already know the order of arguments for BoxShadow. I think it would benefit to either have a record/object represent the shadow's parameters or a labelled function that could optionally be used to allow for named/out-of-order parameters.

**Comment by bryphe** [comment]
@OhadRau - I'd be open to it, and I think it's a great way to 'use the language'. I never remember the order of those parameters, have to look it up every time 😄 

The `BoxShadow.make` seems consistent with how we use Style's at least.

@Akin909 - what do you think?

**Comment by Akin909** [comment]
@bryphe @OhadRau sounds like a great suggestion to make things less error prone, I can tweak what I'm working on to use the `BoxShadow.make` version :+1:

**Comment by bryphe** [comment]
> @bryphe thanks that helped a lot 👍,got another question now I've actually drawn the shadow, I'm not sure I followed the strategy you suggested about using more quads with lessening opacity to draw the blur?

Sweet I just pulled the branch and saw the shadow coming through! 😍 

I actually don't think we'll actually need more quads with this strategy, if we handle the gradients all in the shader with that 'cutoff'. Seems like it'd all be doable in the pixel shader - just need to handle the other directions!

**Comment by bryphe** [comment]
Implemented with #107 - thank you @Akin909 !
